{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# JKSpliceAI-PyTorch Demo\n",
    "\n",
    "This notebook demonstrates how to use the `jkspliceai-pytorch` package to predict splicing effects of variants, supporting both the standard `10k` model and the `10k_drop` model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "# Add parent directory to path to ensure we can import jkspliceai_pytorch\n",
    "sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '../..'))) # Just in case\n",
    "\n",
    "from jkspliceai_pytorch import spliceAI\n",
    "from jklib.genome import locus\n",
    "import torch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check Device\n",
    "Check if CUDA (NVIDIA GPU) or MPS (Apple Silicon) is available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using MPS\n"
     ]
    }
   ],
   "source": [
    "if torch.cuda.is_available():\n",
    "    print(\"Using CUDA\")\n",
    "elif torch.backends.mps.is_available():\n",
    "    print(\"Using MPS\")\n",
    "else:\n",
    "    print(\"Using CPU\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Variant\n",
    "Define the genomic location, reference allele, and alternative allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "loc = locus(\"chr1:925952-925952\")\n",
    "ref = \"G\"\n",
    "alt = \"A\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Run with Standard Model (10k)\n",
    "This uses the default SpliceAI 10k model weights (`models_data/pytorch_10k`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dowonkim/Library/CloudStorage/Dropbox/project/modules/jkspliceai-pytorch/jkspliceai_pytorch/wrapper/runner.py:71: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state_dict = torch.load(\n",
      "/Users/dowonkim/Library/CloudStorage/Dropbox/project/modules/jkspliceai-pytorch/jkspliceai_pytorch/wrapper/runner.py:71: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state_dict = torch.load(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standard Model Results:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pre-mRNA position (Acceptor Loss)</th>\n",
       "      <th>delta score (Acceptor Loss)</th>\n",
       "      <th>pre (Acceptor Loss)</th>\n",
       "      <th>post (Acceptor Loss)</th>\n",
       "      <th>pre-mRNA position (Donor Loss)</th>\n",
       "      <th>delta score (Donor Loss)</th>\n",
       "      <th>pre (Donor Loss)</th>\n",
       "      <th>post (Donor Loss)</th>\n",
       "      <th>pre-mRNA position (Acceptor Gain)</th>\n",
       "      <th>delta score (Acceptor Gain)</th>\n",
       "      <th>pre (Acceptor Gain)</th>\n",
       "      <th>post (Acceptor Gain)</th>\n",
       "      <th>pre-mRNA position (Donor Gain)</th>\n",
       "      <th>delta score (Donor Gain)</th>\n",
       "      <th>pre (Donor Gain)</th>\n",
       "      <th>post (Donor Gain)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-54</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.56</td>\n",
       "      <td>-152</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-54</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.07</td>\n",
       "      <td>61</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-2542</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3172</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-428</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-2330</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-187</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-2416</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-1187</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2105</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pre-mRNA position (Acceptor Loss)  delta score (Acceptor Loss)  \\\n",
       "0                                11                          0.0   \n",
       "1                                -1                          0.0   \n",
       "2                             -2542                          0.0   \n",
       "3                             -2330                          0.0   \n",
       "4                             -2416                          0.0   \n",
       "\n",
       "   pre (Acceptor Loss)  post (Acceptor Loss) pre-mRNA position (Donor Loss)  \\\n",
       "0                 0.00                  0.00                            -54   \n",
       "1                 0.00                  0.00                            -30   \n",
       "2                 0.03                  0.03                            -25   \n",
       "3                 0.00                  0.00                          -1165   \n",
       "4                 0.01                  0.01                          -1187   \n",
       "\n",
       "   delta score (Donor Loss)  pre (Donor Loss)  post (Donor Loss)  \\\n",
       "0                       0.0               0.0                0.0   \n",
       "1                       0.0               0.0                0.0   \n",
       "2                       0.0               0.0                0.0   \n",
       "3                       0.0               0.0                0.0   \n",
       "4                       0.0               0.0                0.0   \n",
       "\n",
       "  pre-mRNA position (Acceptor Gain)  delta score (Acceptor Gain)  \\\n",
       "0                               -30                         0.21   \n",
       "1                               -54                         0.03   \n",
       "2                              3172                         0.01   \n",
       "3                               -75                         0.00   \n",
       "4                              2105                         0.00   \n",
       "\n",
       "   pre (Acceptor Gain)  post (Acceptor Gain) pre-mRNA position (Donor Gain)  \\\n",
       "0                 0.34                  0.56                           -152   \n",
       "1                 0.04                  0.07                             61   \n",
       "2                 0.04                  0.05                           -428   \n",
       "3                 0.01                  0.01                           -187   \n",
       "4                 0.01                  0.02                             10   \n",
       "\n",
       "   delta score (Donor Gain)  pre (Donor Gain)  post (Donor Gain)  \n",
       "0                      0.10              0.27               0.37  \n",
       "1                      0.10              0.79               0.89  \n",
       "2                      0.09              0.25               0.34  \n",
       "3                      0.07              0.13               0.20  \n",
       "4                      0.05              0.05               0.11  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_standard = spliceAI(\n",
    "    loc=loc, \n",
    "    ref=ref, \n",
    "    alt=alt, \n",
    "    model='spliceai_torch', # or '10k'\n",
    "    view=10,\n",
    "    use_gpu=True\n",
    ")\n",
    "\n",
    "print(\"Standard Model Results:\")\n",
    "display(df_standard.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Run with Drop-Layer Model (10k_drop)\n",
    "This uses the `10k_drop` model weights located in `models_data/pytorch`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dowonkim/Library/CloudStorage/Dropbox/project/modules/jkspliceai-pytorch/jkspliceai_pytorch/wrapper/runner.py:71: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state_dict = torch.load(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drop Model Results:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pre-mRNA position (Acceptor Loss)</th>\n",
       "      <th>delta score (Acceptor Loss)</th>\n",
       "      <th>pre (Acceptor Loss)</th>\n",
       "      <th>post (Acceptor Loss)</th>\n",
       "      <th>pre-mRNA position (Donor Loss)</th>\n",
       "      <th>delta score (Donor Loss)</th>\n",
       "      <th>pre (Donor Loss)</th>\n",
       "      <th>post (Donor Loss)</th>\n",
       "      <th>pre-mRNA position (Acceptor Gain)</th>\n",
       "      <th>delta score (Acceptor Gain)</th>\n",
       "      <th>pre (Acceptor Gain)</th>\n",
       "      <th>post (Acceptor Gain)</th>\n",
       "      <th>pre-mRNA position (Donor Gain)</th>\n",
       "      <th>delta score (Donor Gain)</th>\n",
       "      <th>pre (Donor Gain)</th>\n",
       "      <th>post (Donor Gain)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.88</td>\n",
       "      <td>61</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-54</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.35</td>\n",
       "      <td>10</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1138</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1852</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>921</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "      <td>58</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-54</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-75</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-152</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>498</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-2379</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.80</td>\n",
       "      <td>-475</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pre-mRNA position (Acceptor Loss)  delta score (Acceptor Loss)  \\\n",
       "0                                11                          0.0   \n",
       "1                                -1                          0.0   \n",
       "2                              1138                          0.0   \n",
       "3                                 8                          0.0   \n",
       "4                                61                          0.0   \n",
       "\n",
       "   pre (Acceptor Loss)  post (Acceptor Loss) pre-mRNA position (Donor Loss)  \\\n",
       "0                 0.01                   0.0                            999   \n",
       "1                 0.01                   0.0                           1448   \n",
       "2                 0.00                   0.0                           1852   \n",
       "3                 0.00                   0.0                            -54   \n",
       "4                 0.00                   0.0                            498   \n",
       "\n",
       "   delta score (Donor Loss)  pre (Donor Loss)  post (Donor Loss)  \\\n",
       "0                       0.0              0.05               0.04   \n",
       "1                       0.0              0.05               0.05   \n",
       "2                       0.0              0.03               0.02   \n",
       "3                       0.0              0.00               0.00   \n",
       "4                       0.0              0.00               0.00   \n",
       "\n",
       "  pre-mRNA position (Acceptor Gain)  delta score (Acceptor Gain)  \\\n",
       "0                               -30                         0.16   \n",
       "1                               -54                         0.09   \n",
       "2                               921                         0.01   \n",
       "3                               -75                         0.01   \n",
       "4                             -2379                         0.01   \n",
       "\n",
       "   pre (Acceptor Gain)  post (Acceptor Gain) pre-mRNA position (Donor Gain)  \\\n",
       "0                 0.72                  0.88                             61   \n",
       "1                 0.26                  0.35                             10   \n",
       "2                 0.05                  0.06                             58   \n",
       "3                 0.01                  0.02                           -152   \n",
       "4                 0.79                  0.80                           -475   \n",
       "\n",
       "   delta score (Donor Gain)  pre (Donor Gain)  post (Donor Gain)  \n",
       "0                      0.12              0.77               0.89  \n",
       "1                      0.09              0.10               0.19  \n",
       "2                      0.07              0.07               0.14  \n",
       "3                      0.05              0.35               0.40  \n",
       "4                      0.04              0.33               0.37  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_drop = spliceAI(\n",
    "    loc=loc, \n",
    "    ref=ref, \n",
    "    alt=alt, \n",
    "    model='10k_drop',\n",
    "    view=10,\n",
    "    use_gpu=True\n",
    ")\n",
    "\n",
    "print(\"Drop Model Results:\")\n",
    "display(df_drop.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "travel",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
